rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -1024,6 +1024,34 @@ impl SystemState {
         subnet_available_guaranteed_response_memory: &mut i64,
         own_subnet_type: SubnetType,
         input_queue_type: InputQueueType,
+    ) -> Result<bool, (StateError, RequestOrResponse)> {
+        #[cfg(debug_assertions)]
+        let balance_before = self.balance_with_messages(Some(msg.cycles()));
+
+        let res = self.push_input_impl(
+            msg,
+            subnet_available_guaranteed_response_memory,
+            own_subnet_type,
+            input_queue_type,
+        );
+
+        #[cfg(debug_assertions)]
+        self.assert_balance_with_messages(
+            balance_before,
+            res.as_ref().err().map(|(_, msg)| msg.cycles()),
+        );
+
+        res
```

Comment:
This (repeated 4x) is the best I was able to come up with as a way of comparing before and after cycle balances. It will also take into account the "outbound refund pool" in the following PR.

Any better suggestions are welcome.